### ISSUE TITLE
Conditional isinstance should intersect types

### ISSUE THREAD

**Issue by NeilGirdhar** [microsoft/pyright#2812]
```python
from collections.abc import Callable
from typing import Any, Generic, TypeVar

R = TypeVar('R')


class DemoMethod(Generic[R]):
    def __init__(self, method: Callable[[], R]):
        self.method: Callable[[], R] = method.__func__

    def __call__(self, x: bool, **kwargs: Any) -> R:
        reval = self.method(**kwargs)
        reveal_type(self.method)  # "self.method" is "() -> R@DemoMethod"
        if x or isinstance(self, PlotMethod):
            pass
        reveal_type(self.method)  # "self.method" is "(() -> R@DemoMethod) | (() -> None)"
        return self.method()


class PlotMethod(DemoMethod[None]):
    pass
```
Produces:
```
  /home/neil/src/cmm/a.py:17:16 - error: Expression of type "R@DemoMethod | None" cannot be assigned to return type "R@DemoMethod"
    Type "R@DemoMethod | None" cannot be assigned to type "R@DemoMethod"
```

I'm not sure if Pyright supports intersections (internally) yet, but it seems that within the condition, the revealed type should be the intersection of the type going in (that uses the generic) and the type enforced by the condition (that uses None).

**Comment by erictraut** [comment]
Pyright is doing the right thing here. Within the `if` branch, `self` is narrowed to `PlotMethod`. In the "implied else" branch, it remains type `Self`. After the `if` statement it is the union of the two types (PlotMethod and Self), so the expression `self.method` is the union of the two method types. An intersection type would not be appropriate here.

**Comment by NeilGirdhar** [comment]
> Pyright is doing the right thing here. 

Sorry, but I'm fairly sure you are mistaken.

> After the if statement it is the union of the two types (PlotMethod and Self), 

It should be impossible for a condition (that assigns nothing) to lead to a broadening of types.

After the if statement, the type of `self.method` should be "(() -> R@DemoMethod) | (() -> None) & () -> R@DemoMethod"`.  Where `&` represents an intersection.  MyPy gets this wrong too in some places.

> so the expression self.method is the union of the two method types. 

Yes, it's the union of the two types, but the second type in the union must be intersected with the previous constraint involving the type variable.

**Comment by erictraut** [comment]
OK, I understand your argument. I need to think about this more.

My inclination is that this is a very esoteric edge case and probably not worth addressing. If I can think of a straightforward fix, I'll consider it.

The code in the example is using an anti-pattern where a base class has knowledge of a derived class. If you restructure this code to avoid this anti-pattern, you will also avoid this type issue.

**Comment by NeilGirdhar** [comment]
Yes, I agree this is an anti-pattern.  Let me look for the MyPy example of a similar error.

**Comment by NeilGirdhar** [comment]
Here's a couple:
- https://github.com/python/mypy/issues/9424 (pyright fails on this)
- https://github.com/python/mypy/issues/5720 (pyright works on this!)

**Comment by erictraut** [comment]
I think pyright succeeds on both of these, no?

**Comment by NeilGirdhar** [comment]
Yes, it works better than MyPy here.  However, I think I constructed an example that Pyright doesn't work on from the first issue:
```
from typing import Optional, TypeVar

U = TypeVar('U', bound=Optional[int])

def g(x: None) -> None:
    pass

def f(u: U) -> U:
    if u is None:
        g(u) # error: Argument of type "U@f" cannot be assigned to parameter "x" of type "None" in function "g"
        return None  # error: Expression of type "None" cannot be assigned to return type "U@f"
    assert isinstance(u, int)  # removing this line causes it to pass.
    return u
```
In the first case, `u` shoudl have type `None & U`, and so should not cause an error.  The second case is trickier, and I don't expect Pyright to succeed here, although it would technically be possible.

**Comment by erictraut** [comment]
The `isinstance` narrowing logic in pyright is employing a technique that I call "conditional types". You can read about it in [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#constrained-type-variables-and-conditional-types). It's similar to what you're referring to as an intersection type here.

The `x is None` type narrowing logic doesn't currently support conditional types, but it would be easy to add this. That would eliminate the false positive for `g(u)`.

I also wouldn't expect `return None` to work. The appropriate workaround is to change it to `return u`.

In your example at the top of this thread, conditional types don't work because they don't propagate to member access expression and call expression evaluation. I don't think it's feasible to make that work.

**Comment by NeilGirdhar** [comment]
> It's similar to what you're referring to as an intersection type here.

Right, makes perfect sense.  When I replace `NoneType` with a class `C`, and reveal the type before the first return, I get:
```
info: Type of "u" is "<subclass of int and C>*"
```
Very cool!

> I also wouldn't expect return None to work. The appropriate workaround is to change it to return u.

Yes, I agree.

> In your example at the top of this thread, conditional types don't work because they don't propagate to member access expression and call expression evaluation. I don't think it's feasible to make that work.

Okay.  These type checkers are way beyond me.  Thanks for taking a look anyhow.  I just started using Pyright today, and it found a good 20 errors that MyPy missed.

**Comment by erictraut** [comment]
I've updated the `x is None` type guard logic to preserve type conditions, so your example above now works. This will be included in the next release.

As for the example at the top of this thread, I don't see a straightforward way to make that work.

**Comment by NeilGirdhar** [comment]
Okay, thanks again, please feel free to close this.

**Comment by erictraut** [comment]
This is included in pyright 1.1.204, which I just published. It will also be included in the next release of pylance.
